python/bindings.go

```diff
@@ -0,0 +1,389 @@
+    siadClient := renterutil.NewSiadClient(siadAddr, siadPassword)
+    return storePtr(siadClient)
+}
+
+//export us_ll_form_contract
+func us_ll_form_contract(id unsafe.Pointer, client_p unsafe.Pointer, host_str *C.char, key_ptr unsafe.Pointer, total_funds *C.char, duration C.uint) unsafe.Pointer {
+    clientMu.Lock()
+    defer clientMu.Unlock()
+    siad := loadPtr(client_p).(*renterutil.SiadClient)
+    hostKeyPrefix := C.GoString(host_str)
+    totalFunds := C.GoString(total_funds)
+
+    hostKey, err := siad.LookupHost(hostKeyPrefix)
+    if setError(id, err) {
+        return nil
+    }
+    addr, err := siad.ResolveHostKey(hostKey)
+    if setError(id, err) {
+        return nil
+    }
+    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+    defer cancel()
+    host, err := hostdb.Scan(ctx, addr, hostKey)
+    if setError(id, err) {
+        return nil
+    }
+
+    currentHeight, err := siad.ChainHeight()
+    if setError(id, err) {
+        return nil
+    }
+
+    key := ed25519.NewKeyFromSeed(goBytes(key_ptr, 32))
+    funds, err := scanCurrency(totalFunds)
+    if setError(id, err) {
+        return nil
+    }
+    contract, _, err := proto.FormContract(siad, siad, key, host, funds, currentHeight, currentHeight+types.BlockHeight(duration))
+    if setError(id, err) {
+        return nil
+    }
+    buf := make([]byte, C.sizeof_struct_contract_t)
+    copy(buf[0:32], contract.HostKey().Ed25519())
+    copy(buf[32:64], contract.Revision.ParentID[:])
+    copy(buf[64:96], key[:ed25519.SeedSize])
+    return C.CBytes(buf)
+}
+
+//export us_ll_new_session
+func us_ll_new_session(id unsafe.Pointer, client_p unsafe.Pointer, host_str *C.char, contract *C.struct_contract_t) unsafe.Pointer {
+    clientMu.Lock()
+    defer clientMu.Unlock()
+    siad := loadPtr(client_p).(*renterutil.SiadClient)
+    hostKeyPrefix := C.GoString(host_str)
+
+    hostKey, err := siad.LookupHost(hostKeyPrefix)
+    if setError(id, err) {
+        return nil
+    }
+    addr, err := siad.ResolveHostKey(hostKey)
+    if setError(id, err) {
+        return nil
+    }
+    currentHeight, err := siad.ChainHeight()
+    if setError(id, err) {
+        return nil
+    }
+
+    var c renter.Contract
+    copy(c.ID[:], C.GoBytes(unsafe.Pointer(&contract.id), 32))
+    c.HostKey = hostdb.HostKeyFromPublicKey(C.GoBytes(unsafe.Pointer(&contract.hostKey), 32))
+    c.RenterKey = ed25519.NewKeyFromSeed(C.GoBytes(unsafe.Pointer(&contract.renterKey), 32))
+    session, err := proto.NewSession(addr, hostKey, c.ID, c.RenterKey, currentHeight)
+    if setError(id, err) {
+        return nil
+    }
+    return storePtr(session)
+}
+
+//export us_ll_upload
+func us_ll_upload(id unsafe.Pointer, session_p unsafe.Pointer, buf unsafe.Pointer) unsafe.Pointer {
+    session := loadPtr(session_p).(*proto.Session)
+    var sector [renterhost.SectorSize]byte
+    copy(sector[:], goBytes(buf, renterhost.SectorSize))
+    root, err := session.Append(&sector)
```

Comment:
you could probably save a copy here by casting `buf` to a  `*[renterhost.SectorSize]byte` directly. That's a little dangerous, though. You want to have some protection against the caller passing a "sector" with the wrong size.

Fine for now, though.